### PR TITLE
Fixing issue with uniontypes not created with primitive types

### LINKED ISSUE #9531: webidl uniontype dont generate for StringOrUnsignedLong
I would like to ask for some help understanding this problem with the webidl-s
[bluetooth.rs](https://gist.github.com/zakorgy/e4d235b198cbc6182c7f)
[Bluetooth.webidl](https://gist.github.com/zakorgy/f7e35b80f74bbb76784e)
_the code line in the webidl file:_
`static UUID getService((DOMString or unsigned long) name)
`
_at first build attempt:_
```
AssertionError
make: *** [/servo/target/release/build/script-ccd19d3c812c83f8/out/ParserResults.pkl] Error 1
thread '<main>' panicked at 'assertion failed: Command::new("make").args(&["-f",
 "makefile.cargo"]).status().unwrap().success()', /servo/components/script/build.rs:8
```
_at second attempt this is what i expected:_
`error use of undeclared type name UnionTypes::StringOrUnsignedLong`

_but with the import (dom::bindings::codegen::UnionTypes::StringOrUnsignedLong ):_
`error: unresolved import dom::bindings::codegen::UnionTypes::StringOrUnsignedLong. There is no StringOrUnsignedLong in dom::bindings::codegen::UnionT`

### PR DESCRIPTION
**Pull request by zakorgy**
Fixes #9531

[Review on Reviewable](https://reviewable.io/reviews/servo/servo/9580)

### CODE REVIEW

**Comment by jdm** [comment]
Please do add a case to `TestBinding.webidl` that exercises these changes!

File: components/script/dom/bindings/codegen/CodegenRust.py [hunk 1]
```diff
@@ -3757,7 +3755,35 @@ def get_match(name):
             match = (
                 "match %s::TryConvertTo%s(cx, value) {\n"
                 "    Err(_) => return Err(()),\n"
-                "    Ok(Some(value)) => return Ok(%s::%s(value)),\n"
+                "    Ok(Some(value)) => return Ok(%s::e%s(value)),\n"
```
Comment (jdm) [inline components/script/dom/bindings/codegen/CodegenRust.py]: The `e` prefix should be removed.

File: components/script/dom/bindings/codegen/CodegenRust.py [hunk 2]
```diff
@@ -3757,7 +3755,35 @@ def get_match(name):
...
+
+        numericTypes = [t for t in memberTypes if t.isNumeric()]
+        if len(numericTypes) > 0:
+            assert len(numericTypes) == 1
+            memberType = numericTypes[0]
+            name = memberType.name
+            match = (
+                "match %s::TryConvertTo%s(cx, value) {\n"
+                "    Err(_) => return Err(()),\n"
+                "    Ok(Some(value)) => return Ok(%s::e%s(value)),\n"
```
Comment (jdm) [inline components/script/dom/bindings/codegen/CodegenRust.py]: Remove the `e` prefix.

File: components/script/dom/bindings/codegen/CodegenRust.py [hunk 3]
```diff
@@ -3757,7 +3755,35 @@ def get_match(name):
...
+
+        booleanTypes = [t for t in memberTypes if t.isBoolean()]        
+        if len(booleanTypes) > 0:
+            assert len(booleanTypes) == 1
+            memberType = booleanTypes[0]
+            name = memberType.name
+            match = (
+                "match %s::TryConvertTo%s(cx, value) {\n"
+                "    Err(_) => return Err(()),\n"
+                "    Ok(Some(value)) => return Ok(%s::e%s(value)),\n"
```
Comment (jdm) [inline components/script/dom/bindings/codegen/CodegenRust.py]: Remove the `e` prefix.

**Comment by jdm** [comment]
Just pointing out a couple issues that the Travis build noticed. Be sure to run `./mach test-tidy` to catch style-related issues. Thanks for working on this!

**Comment by zakorgy** [comment]
Thx for fast review

**Comment by Ms2ger** [comment]
There's a reason the Gecko code I quoted uses the strange
```python
        if stringTypes or numericTypes or booleanTypes:
            if stringConversion:
                if booleanConversion:
```
pattern: for `(DOMString or unsigned long)`, this code will never create an `UnsignedLong` variant. Note step 13 and 14 of <https://heycam.github.io/webidl/#es-union>.

**Comment by Ms2ger** [comment]
Could you please rebase and squash? We can't merge this PR if it contains merge commits.

File: components/script/dom/bindings/codegen/CodegenRust.py [hunk 4]
```diff
@@ -3744,24 +3744,56 @@ def get_match(name):
...
+                assert booleanConversion
+                other.append(booleanConversion[0])
+            if hasObjectTypes:
+                if object:
+                    conversions.append(CGWrapper(CGList(other, "\n\n"), pre="else {\n", post="\n}"))
+                else:
+                    conversions.append(CGWrapper(CGList(other, "\n\n"), pre="", post=""))
+            else:
+                conversions.append(CGWrapper(CGList(other, "\n\n"), pre="", post=""))
+        conversions.append(CGGeneric(
```
Comment (Ms2ger) [inline components/script/dom/bindings/codegen/CodegenRust.py]: This code seems to be duplicated

File: components/script/dom/bindings/codegen/CodegenRust.py [hunk 5]
```diff
@@ -3744,24 +3744,56 @@ def get_match(name):
...
-                "    Err(_) => return Err(()),\n"
-                "    Ok(Some(value)) => return Ok(%s::%s(value)),\n"
-                "    Ok(None) => (),\n"
-                "}\n") % (self.type, name, self.type, name)
-            conversions.append(CGGeneric(match))
-            names.append(name)
+                assert booleanConversion
+                other.append(booleanConversion[0])
+            if hasObjectTypes:
+                if object:
```
Comment (Ms2ger) [inline components/script/dom/bindings/codegen/CodegenRust.py]: We don't support `object` being non-null here; just drop the conditionals and use
```python
conversions.append(CGList(other, "\n\n"))
```

**Comment by Ms2ger** [comment]
LGTM with those two comments fixed

**Pull request merged**